lib/seahorn/BvOpSem2RawMemMgr.hh

```diff
@@ -22,71 +25,141 @@ class RawMemManager : public OpSemMemManager {
-
 public:
-  RawMemManager(Bv2OpSem &sem, Bv2OpSemContext &ctx, unsigned int ptrSz,
-                unsigned int wordSz, bool useLambdas);
+  RawMemManagerCore(Bv2OpSem &sem, Bv2OpSemContext &ctx, unsigned int ptrSz,
+                    unsigned int wordSz, bool useLambdas);
 
-  RawMemManager(Bv2OpSem &sem, Bv2OpSemContext &ctx, unsigned ptrSz,
-                unsigned wordSz, bool useLambdas, bool ignoreAlignment);
+  RawMemManagerCore(Bv2OpSem &sem, Bv2OpSemContext &ctx, unsigned ptrSz,
+                    unsigned wordSz, bool useLambdas, bool ignoreAlignment);
 
-  ~RawMemManager() override = default;
+  ~RawMemManagerCore();
 
   OpSemAllocator &getMAllocator() const;
 
   bool ignoreAlignment() const;
 
+  struct MemValTyImpl {
+    Expr m_v;
+
+    explicit MemValTyImpl(Expr &&raw_val) {
+      assert(!raw_val || !strct::isStructVal(raw_val));
+      m_v = std::move(raw_val);
+    }
+
+    explicit MemValTyImpl(const Expr &raw_val) {
+      assert(!raw_val || !strct::isStructVal(raw_val));
+      m_v = raw_val;
+    }
+
+    Expr v() const { return m_v; }
+    Expr toExpr() const { return v(); }
+    explicit operator Expr() const { return toExpr(); }
+
+    Expr getRaw() { return m_v; }
+  };
+
+  struct PtrTyImpl {
+    Expr p_v;
+
+    explicit PtrTyImpl(Expr &&raw_val) {
+      assert(!raw_val || !strct::isStructVal(raw_val));
+      p_v = std::move(raw_val);
+    }
+
+    explicit PtrTyImpl(const Expr &raw_val) {
+      assert(!raw_val || !strct::isStructVal(raw_val));
+      p_v = raw_val;
+    }
+
+    Expr v() const { return p_v; }
+    Expr toExpr() const { return v(); }
+    explicit operator Expr() const { return toExpr(); }
+
+    Expr getRaw() { return p_v; }
+  };
+
+  struct MemSortTyImpl {
+    Expr m_mem_sort;
+
+    explicit MemSortTyImpl(Expr &&mem_sort) {
+      m_mem_sort = std::move(mem_sort);
+    }
+
+    explicit MemSortTyImpl(const Expr &mem_sort) { m_mem_sort = mem_sort; }
+
+    Expr v() const { return m_mem_sort; }
+    Expr toExpr() const { return v(); }
+    explicit operator Expr() const { return toExpr(); }
+  };
+
+  struct PtrSortTyImpl {
+    Expr m_ptr_sort;
+
+    explicit PtrSortTyImpl(Expr &&ptr_sort) {
+      m_ptr_sort = std::move(ptr_sort);
+    }
+
+    explicit PtrSortTyImpl(const Expr &ptr_sort) { m_ptr_sort = ptr_sort; }
+
+    Expr v() const { return m_ptr_sort; }
+    Expr toExpr() const { return v(); }
+    explicit operator Expr() const { return toExpr(); }
+  };
+
   /// Right now everything is an expression. In the future, we might have
   /// other types for PtrTy, such as a tuple of expressions
-  using PtrTy = OpSemMemManager::PtrTy;
-  using MemValTy = OpSemMemManager::MemValTy;
-  using PtrSortTy = OpSemMemManager::PtrSortTy;
-  using MemSortTy = OpSemMemManager::MemSortTy;
-  using MemRegTy = OpSemMemManager::MemRegTy;
+  using PtrTy = PtrTyImpl;
+  using MemValTy = MemValTyImpl;
+  using PtrSortTy = PtrSortTyImpl;
+  using MemSortTy = MemSortTyImpl;
+  using MemRegTy = Expr;
```

Comment:
still an `Expr`. Is that an oversight?

Reply:
This doesn't use `OpSemMemManager` anymore so I didn't want to depend on types exported by that class. Since it was just redefining type, I took the actual definition directly. 

In the future, maybe we can have a class which encapsulates these types.